### ISSUE TITLE
is_action_pressed sometimes skips key press

### ISSUE THREAD

**Issue by OlegDydy** [godotengine/godot#36396]
**Godot version:**
3.2-stable

**OS/device including version:**
Windows 10 Home (Version: 1903, Build 18362.657)

**Issue description:**
Sometimes in-game keystroke events are skipped. This works if you press them at the same time (as much as possible of course). In my case, this happened  with the “X” and arrow buttons. But I think there are many other cases when this can happen.

**Steps to reproduce:**
To reproduce this problem with a 99% guarantee, you will need a gaming mouse/keyboard that can record macros. Or software that can simulate keystrokes (the same macros). Of course, you can do this manually, but it may require some practice.

**My thoughts on this issue:**
I downloaded the source code of the engine and did a little research. I think the problem is in the WM_KEYPRESS event (most likely on the Windows side).
As I understand it, when the OS receives several keystroke events at the same time, it processes them in the order of their virtual key codes, but returns in the order they were received. Therefore, when we extract the first key (eg X) from the event queue, the arrow keys are already processed. Thus, in the WM_KEYPRESS event, the 30th bit of lParam indicates that the arrow key was previously pressed (but the engine did not know about it).
To fix this problem, it probably makes sense to change the [1183 line](https://github.com/godotengine/godot/blob/851cb429631168369d2a51812de763b658795fbf/platform/windows/os_windows.cpp#L1183) of `os_windows.cpp` to something like this:
```c++
k->set_echo((ke.uMsg == WM_KEYDOWN && keyboard_state[ke.wParam]));
keyboard_state[ke.wParam] = ke.uMsg == WM_KEYDOWN;
```
PS keyboard_state is array or std::bitset of 256 values ​​(total amount of virtual key codes).
PSS By the way get_action_strength() works perfectly

**Comment by nathanwfranke** [comment]
Cannot reproduce this on Windows 10 Pro 64 bit, (10.0, Build 18362)

Test project for those who need it:
[Test.zip](https://github.com/godotengine/godot/files/4235338/Test.zip)

```swift
var keys_pressed = 0

func _process(_delta):
	print(keys_pressed)
	if Input.is_action_just_pressed("ui_up"):
		keys_pressed += 1
	if Input.is_action_just_pressed("ui_down"):
		keys_pressed += 1
	if Input.is_action_just_pressed("ui_left"):
		keys_pressed += 1
	if Input.is_action_just_pressed("ui_right"):
		keys_pressed += 1
	if Input.is_action_just_pressed("x"):
		keys_pressed += 1

func _input(event):
	if event.is_action_pressed("ui_up"):
		keys_pressed += 1
	if event.is_action_pressed("ui_down"):
		keys_pressed += 1
	if event.is_action_pressed("ui_left"):
		keys_pressed += 1
	if event.is_action_pressed("ui_right"):
		keys_pressed += 1
	if event.is_action_pressed("x"):
		keys_pressed += 1
```

It prints out the total number of keys you pressed, so just keep pressing all 5 at once and make sure the number is divisible by 10 (2 checks of each type)

[(Serious) Easy way to press all the arrow keys simultaneously. I had quite a bit of trouble with it.](https://imgur.com/a/11rpDFO)

**Comment by Calinou** [comment]
@OlegDydy Does this occur if you disable input accumulation by calling `Input.set_use_accumulated_input(false)` in `_ready()`?

**Comment by OlegDydy** [comment]
@Calinou No it does not. I don't use `Input.set_use_accumulated_input(false)` anywhere. The very first time I encountered this problem was when I figured out the standard Godot demo “2D Platformer Demo (kinematicbody)”

**Comment by OlegDydy** [comment]
@nathanwfranke I think this version ([KeyboardTest.zip](https://github.com/godotengine/godot/files/4236928/KeyboardTest.zip)) of the test demo will be much more informative (and yes, you do not need to press “X” and **all** the arrow keys, just “X” and **any** arrow key)
```GDScript
extends Node

var keys_pressed = ""

func _process(_delta):
	keys_pressed = ""
	keys_pressed += "x" if Input.is_action_pressed("x") else "_"
	keys_pressed += "^" if Input.is_action_pressed("ui_up") else "_"
	keys_pressed += "V" if Input.is_action_pressed("ui_down") else "_"
	keys_pressed += "<" if Input.is_action_pressed("ui_left") else "_"
	keys_pressed += ">" if Input.is_action_pressed("ui_right") else "_"
	print(keys_pressed)
```
If everything is ok, you'll see something like this `x__<_` or this `x^___`
If not, you'll only see `x____`
PS Using your method, I achieved a result when almost all the arrow keys were skipped (the demo printed `x^___` despite the five keys pressed on the keyboard).

**Comment by dreamsComeTrue** [comment]
I've opened your project in current 'master' build. I've been pressing all buttons in random order - and as you see - all of them are being recorded properly.
![demo](https://user-images.githubusercontent.com/1110337/75052505-014d5f00-54d0-11ea-8a86-8dc2e748854d.gif)

**Comment by OlegDydy** [comment]
Need to press keys (eg <X> + <Arrow up>) at the same time (not hold X and **then** press arrows, this case have no problems).

**Comment by dreamsComeTrue** [comment]
Nope. Pressed X and arrow (and couple of them) at the exact same time. I've been also holding them for longer period to find if they start to flicker - but nothing funky happened - I've always had: x<^ in the output. Maybe it's something with your keybaord config (like key-strokes ratio) on Windows.
Does that error occur on other OS? Like, did you check it on Linux/macOS ?

**Comment by nathanwfranke** [comment]
Anyone who can reproduce this, make sure you have an anti-ghosting keyboard and no issues within keyboard testing apps.

**Comment by OlegDydy** [comment]
@dreamsComeTrue I ran demo on Linux (on the same machine) and everything was fine. This only happens on Windows. Of course, I do not exclude the possibility of something wrong with my Windows installation, and not with Godot.

**Comment by OlegDydy** [comment]
A small update. This error seems to depend on the frame rate. When I turn off VSync and make the frame rate unlimited, the problem disappears. And when I set the frame rate, for example, to 1 frame per second, it becomes more noticeable. For some reason key down event enters the program with `is_echo()` equals to true. In attached image the output from `_unhandled_input` callback.
![изображение](https://user-images.githubusercontent.com/30150630/79660466-b0e73a80-81bc-11ea-9327-0685a82f5582.png)

**Comment by WebF0x** [comment]
To add on @OlegDydy, here's a snippet to consistently reproduce the bug
```
extends Node

var should_be_zero_when_spacebar_released = 0

func _ready():
	assert(
		ProjectSettings["application/run/frame_delay_msec"] == 1000,
		"Set your frame rate to 1000 in Project > Project Settings > General > Application > Run > Frame Delay Msec")
	print("Tap spacebar between the frames/prints. If the counter increases, you reproduced the bug")
func _process(_delta):
	if Input.is_action_just_pressed("ui_accept"):
		should_be_zero_when_spacebar_released -= 1
	if Input.is_action_just_released("ui_accept"):
		should_be_zero_when_spacebar_released += 1
	print(should_be_zero_when_spacebar_released)

```

Output:
```
Tap spacebar between the frames/prints. If the counter increases, you reproduced the bug
0
0
1
2
3
4
5
5
```
